src/Processors/Formats/Impl/FreeformRowInputFormat.cpp

```diff
@@ -0,0 +1,473 @@
+#include "Processors/Formats/Impl/FreeformRowInputFormat.h"
+#include <DataTypes/DataTypeString.h>
+#include <IO/ReadBufferFromString.h>
+#include "Common/Exception.h"
+#include "Common/StringUtils/StringUtils.h"
+#include <Common/logger_useful.h>
+#include "Core/NamesAndTypes.h"
+#include "DataTypes/DataTypeNullable.h"
+#include "DataTypes/DataTypesNumber.h"
+#include "DataTypes/IDataType.h"
+#include "DataTypes/Serializations/ISerialization.h"
+#include "Formats/EscapingRuleUtils.h"
+#include "Formats/FormatFactory.h"
+#include "Formats/JSONUtils.h"
+#include "IO/ReadHelpers.h"
+#include "Processors/Formats/IRowInputFormat.h"
+#include "Processors/Formats/ISchemaReader.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int CANNOT_READ_ALL_DATA;
+    extern const int UNSUPPORTED_METHOD;
+}
+
+static inline void skipWhitespacesAndDelimiters(ReadBuffer & in)
+{
+    while (!in.eof() && (isWhitespaceASCII(*in.position()) || *in.position() == ',' || *in.position() == ':'))
+        ++in.position();
+}
+
+// Returns the score of the given type. This doesn't take nullable into account.
+// Possible return values (increasing by power of 5):
+// - 1: String
+// - 5: Decimal, Float
+// - 25: Map, Array, Tuple
+// - 125: Date, DateTime,
+static size_t scoreForType(const DataTypePtr & type)
+{
+    WhichDataType which(type);
+
+    if (which.isNullable())
+    {
+        const auto * nullable_type = assert_cast<const DataTypeNullable *>(type.get());
+        return scoreForType(nullable_type->getNestedType());
+    }
+
+    if (which.isDateOrDate32() || which.isDateTimeOrDateTime64())
+        return 125;
+
+    if (which.isMap() || which.isArray() || which.isTuple())
+        return 25;
+
+    if (which.isDecimal() || which.isFloat() || which.isInt() || which.isUInt())
+        return 5;
+
+    return 1;
+}
+
+static size_t scoreForRule(FormatSettings::EscapingRule rule)
+{
+    switch (rule)
+    {
+        case FormatSettings::EscapingRule::JSON:
+            return 4;
+        case FormatSettings::EscapingRule::CSV:
+            [[fallthrough]];
+        case FormatSettings::EscapingRule::Quoted:
+            return 2;
+        case FormatSettings::EscapingRule::Escaped:
+            [[fallthrough]];
+        case FormatSettings::EscapingRule::Raw:
+            return 1;
+        default:
+            return 0;
+    }
+    UNREACHABLE();
+}
+
+static size_t scoreForField(FormatSettings::EscapingRule rule, const DataTypePtr & type)
+{
+    return scoreForRule(rule) * scoreForType(type);
+}
+
+std::vector<std::pair<String, String>> JSONFieldMatcher::parseFields(ReadBuffer & in, size_t index) const
+{
+    if (in.eof())
+        return {};
+
+    if (*in.position() != '{')
+    {
+        String ret;
+        readJSONField(ret, in);
+        return {{fmt::format("c{}", index), ret}};
+    }
+
+
+    ++in.position();
```

Comment:
Do I understand correctly, that here we are trying to parse a JSON object with multiple fields as separate columns like in JSONEachRow?
I think here we can read only valid JSON objects here and don't try to parse it with some heuristics (as it's done now). For example similar to function `readRowAndGetNamesAndDataTypesForJSONEachRow` from `src/Formats/JSONUtils.cpp`, but in our case we should return vector of pairs <FieldName, FieldValue>. We can write a seperate function for it.